references/common/metrics.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2022 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"context"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+	"k8s.io/metrics/pkg/apis/metrics/v1beta1"
+	metricsclientset "k8s.io/metrics/pkg/client/clientset/versioned"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	appv1beta1 "github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	"github.com/oam-dev/kubevela/pkg/velaql/providers/query"
+)
+
+// ApplicationMetrics is the metrics of application
+type ApplicationMetrics struct {
+	Status   *ApplicationMetricsStatus
+	Resource *ApplicationResourceStatus
+}
+
+// ApplicationMetricsStatus is the status of application metrics
+type ApplicationMetricsStatus struct {
+	CPU     uint64
+	Memory  uint64
+	Storage uint64
+}
+
+// ApplicationResourceStatus is the status of application resource
+type ApplicationResourceStatus struct {
+	NodeNum        int
+	ClusterNum     int
+	SubresourceNum int
+	PodNum         int
+	ContainerNum   int
```

Comment:
The cluster may be controlled by ServiceMesh(Istio), and ServiceMesh may inject some sidecar containers automatically. Does `ContainerNum` count them?
If yes, may user feel confused? because it is not specified by user in Application.

Reply:
I think some questions can be answer at same time. Now we use `*v1beta1.ResourceTracker` object to track application's resource. So, if resource belong to an application, it can be tracked out. 

I am not very clear about the logic of ResourceTracker, but from my perspective I think these sidecar containers you mentioned won't be trcaked.